File: src/Javascript/dynamic programming/53.最大子序和.js
/*
 * @lc app=leetcode.cn id=53 lang=javascript
 *
 * [53] 最大子序和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let maxSum = nums[0], pre = 0;
    for (let i = 0; i < nums.length; i++) {
        // update current dp
        pre = Math.max(pre + nums[i], nums[i]);
        // update max value
        maxSum = Math.max(maxSum, pre);
    }
    return maxSum;
};
// @lc code=end

File: src/Java/Mirosoft/53.最大子序和.java
/*
 * @lc app=leetcode.cn id=53 lang=java
 *
 * [53] 最大子序和
 */

// @lc code=start
class Solution {
    public int maxSubArray(int[] nums) {
        int maxSum = nums[0], pre = nums[0];
        for (int i = 1; i < nums.length; i++) {
            pre = Math.max(nums[i], pre + nums[i]);
            maxSum = Math.max(maxSum, pre);
        }
        return maxSum;
    }
}
// @lc code=end

